driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -274,7 +290,7 @@ public static String quote(String id) {
      * CQL keyword, {@code false} otherwise.
      */
     public static boolean isReservedCqlKeyword(String id) {
-        return id != null && RESERVED_KEYWORDS.contains(id.toLowerCase());
+        return id != null && RESERVED_KEYWORDS.contains(id);
```

Comment:
Unfortunately this won't work for mixed-case keywords, as in `Local_Quorum`. I know this is a silly edge case, but it's possible so we have to handle it all the way.

Reply:
yes, agree, missed the case, will think how is better to deal with it

Reply:
To expand on my suggestion in the review summary, we can try using this method:
```java
private static int caseInsensitiveHash(String s) {
    int hash = 17;
    for (int i = 0; i < s.length(); i++) {
        char c = s.charAt(i);
        if (c >= 65 && c <= 90) { // A-Z
            c += 32;
        }
        hash = 31 * hash + c;
    }
    return hash;
```
We turn `RESERVED_KEYWORDS` into a map of caseInsensitiveHash => lower-case keyword.
Then `isReservedKeywords` can be implemented as:
```java
if (id == null) {
  return false;
}
int hash = caseInsensitiveHash(id);
String keyword = RESERVED_KEYWORDS.get(hash);
return keyword != null && id.toLowerCase().equals(keyword);
```
Again I'm not sure if this will improve performance so we'll have to benchmark it.

Reply:
Thank you for the details, I got the idea. Checking different options with JMH, will publish results soon.

Reply:
Some initial results:

```
# JMH version: 1.19
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: C:\Program Files\Java\jre1.8.0_151\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 30 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 2 threads, will synchronize iterations
# Benchmark mode: Average time, time/op

Benchmark                                                                      Mode  Cnt    Score   Error  Units
MetadataBenchmark.equalsWrapper_NonReservedKeywordMixedCase                    avgt   30   35.298 ± 1.840  ns/op
MetadataBenchmark.equalsWrapper_NonReservedKeywordUpperCase                    avgt   30   39.861 ± 0.864  ns/op
MetadataBenchmark.equalsWrapper_ReservedKeywordMixedCase                       avgt   30   66.610 ± 1.391  ns/op
MetadataBenchmark.equalsWrapper_reservedKeywordUpperCase                       avgt   30   92.498 ± 1.685  ns/op
MetadataBenchmark.fullyUpperOrFullyLowerCaseCheck_NonReservedKeywordMixedCase  avgt   30   40.995 ± 0.725  ns/op
MetadataBenchmark.fullyUpperOrFullyLowerCaseCheck_NonReservedKeywordUpperCase  avgt   30   30.344 ± 0.538  ns/op
MetadataBenchmark.fullyUpperOrFullyLowerCaseCheck_ReservedKeywordMixedCase     avgt   30   70.420 ± 3.180  ns/op
MetadataBenchmark.fullyUpperOrFullyLowerCaseCheck_ReservedKeywordUpperCase     avgt   30   20.327 ± 0.635  ns/op
MetadataBenchmark.toLowCase_NonReservedKeywordMixedCase                        avgt   30  130.456 ± 2.344  ns/op
MetadataBenchmark.toLowCase_NonReservedKeywordUpperCase                        avgt   30  161.837 ± 2.163  ns/op
MetadataBenchmark.toLowCase_ReservedKeywordMixedCase                           avgt   30  128.490 ± 4.897  ns/op
MetadataBenchmark.toLowCase_reservedKeywordUpperCase                           avgt   30  142.862 ± 2.836  ns/op
```

Where:

1) equalsWrapper - wrap String with case-insensitive equals and hashcode
```java
private static class IgnoreCaseWrapper {
    private final String str;

    private IgnoreCaseWrapper(String str) {
        this.str = str;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof IgnoreCaseWrapper)) return false;
        IgnoreCaseWrapper anotherWrapper = (IgnoreCaseWrapper) o;

        if (str.length() != anotherWrapper.str.length()) return false;

        String anotherString = ((IgnoreCaseWrapper) o).str;

        for (int i = 0; i < str.length(); i++) {
            char c1 = str.charAt(i);
            char c2 = anotherString.charAt(i);
            if (c1 == c2) {
                continue;
            }
            char low1 = toLowerCaseAscii(c1);
            char low2 = toLowerCaseAscii(c2);
            if (low1 == low2) {
                continue;
            }
            return false;
        }
        return true;
    }

    private static char toLowerCaseAscii(char c) {
        if (c >= 65 && c <= 90) { // A-Z
            c += 32; // convert to low case
        }
        return c;
    }

    @Override
    public int hashCode() {
        int hashCode = 17;
        for (int i = 0; i < str.length(); i++) {
            char c = toLowerCaseAscii(str.charAt(i));
            hashCode = 31 * hashCode + c;
        }
        return hashCode;
    }
}
```java

2) fullyUpperOrFullyLowerCaseCheck - check check is String contains only Upper case or only Lower case symbols and use plain set check in this case, use equalsWrapper otherwise.
```java
public static boolean isReservedCqlKeyword(String id) {
    if(id == null) return false;
    if (isFullyUpperOrFullyLowerCase(id)) {
        return RESERVED_KEYWORDS.contains(id);
    }
    return RESERVED_KEYWORDS_WITH_CUSTOM_EQUAL.contains(ignoreCase(id));
}

private static boolean isFullyUpperOrFullyLowerCase(String s) {
    boolean isAlphanumericLowCase = true;
    boolean isAlphanumericUpperCase = true;
    for (int i = 0; i < s.length(); i++) {
        char c = s.charAt(i);
        if (c >= 65 && c <= 90) { // A-Z
            isAlphanumericLowCase = false;
            if (!isAlphanumericUpperCase) {
                break;
            }
        } else if (c >= 97 && c <= 122) { // a-z
            isAlphanumericUpperCase = false;
            if (!isAlphanumericLowCase) {
                break;
            }
        } else if (!(
                (c >= 48 && c <= 57) // 0-9
                        || (c == 95) // _ (underscore)
        )) {
            isAlphanumericLowCase = false;
            isAlphanumericUpperCase = false;
            break;
        }
    }

    return isAlphanumericLowCase || isAlphanumericUpperCase;
}
```
3) toLowCase - the current behavior with toLowerCase conversion
```java
public static boolean isReservedCqlKeyword(String id) {
    return id != null && RESERVED_KEYWORDS_OLD.contains(id.toLowerCase());
}
```